wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/watcher/FileTreeWalker.java

```diff
@@ -0,0 +1,175 @@
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.vfs.watcher;
+
+import com.google.inject.Inject;
+
+import org.eclipse.che.commons.schedule.executor.ThreadPullLauncher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import static java.nio.file.FileVisitResult.CONTINUE;
+import static java.nio.file.FileVisitResult.SKIP_SUBTREE;
+import static java.nio.file.Files.exists;
+import static java.nio.file.Files.walkFileTree;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.stream.Collectors.toSet;
+
+@Singleton
+public class FileTreeWalker {
+    private static final Logger LOG = LoggerFactory.getLogger(FileTreeWalker.class);
+
+    private final File root;
+
+    private final Set<Consumer<Path>> directoryUpdateConsumers;
+    private final Set<Consumer<Path>> directoryCreateConsumers;
+    private final Set<Consumer<Path>> directoryDeleteConsumers;
+    private final Set<PathMatcher>    directoryExcludes;
+
+    private final Set<Consumer<Path>> fileUpdateConsumers;
+    private final Set<Consumer<Path>> fileCreateConsumers;
+    private final Set<Consumer<Path>> fileDeleteConsumers;
+    private final Set<PathMatcher>    fileExcludes;
+
+    private final ThreadPullLauncher launcher;
+
+    private final Map<Path, Long> files       = new HashMap<>();
+    private final Map<Path, Long> directories = new HashMap<>();
+
+    @Inject
+    public FileTreeWalker(@Named("che.user.workspaces.storage") File root,
+
+                          @Named("che.fs.directory.update") Set<Consumer<Path>> directoryUpdateConsumers,
+                          @Named("che.fs.directory.create") Set<Consumer<Path>> directoryCreateConsumers,
+                          @Named("che.fs.directory.delete") Set<Consumer<Path>> directoryDeleteConsumers,
+                          @Named("che.fs.directory.excludes") Set<PathMatcher> directoryExcludes,
+
+                          @Named("che.fs.file.update") Set<Consumer<Path>> fileUpdateConsumers,
+                          @Named("che.fs.file.create") Set<Consumer<Path>> fileCreateConsumers,
+                          @Named("che.fs.file.delete") Set<Consumer<Path>> fileDeleteConsumers,
+                          @Named("che.fs.file.excludes") Set<PathMatcher> fileExcludes,
+
+                          ThreadPullLauncher launcher) {
```

Comment:
what is the point of explicit use of ThreadPullLauncher here. Why don't you use simple schedule annotated method?